add: an example on GauGAN

sayakpaul opened the pull request:
This example implements the [GauGAN](https://arxiv.org/abs/1903.07291) model for conditional realistic image generation. Following the success of **GauGAN2** we believe this example will be helpful for the community dive deeper into generative deep learning. 

A Colab Notebook for the example is available here: https://colab.research.google.com/drive/18Ar6z6N78nECcm0ftfnJQhYj6jdmxhiy?usp=sharing. 

Cc: @soumik12345

fchollet commented:
Thank you for the PR! Love the topic. I'm currently travelling, but I'll be able to review in 2-3 days.

sayakpaul commented:
No worries, @fchollet. We wanted to end the year with a fun example but we definitely understand your situation.

innat commented:
@sayakpaul 
I think `def build_generator` has an `Input` layer class that didn't import (in the above colab and also in the PR files.)

```python
def build_generator(mask_shape, latent_dim=256):
    latent = Input(shape=(latent_dim))
    mask = Input(shape=mask_shape)

````

soumik12345 commented:
@innat Thanks for pointing out the bug. I have fixed it.

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+        image,
+        segmentation_map,
+        labels,
+        crop_size=tf.convert_to_tensor((IMG_HEIGHT, IMG_WIDTH)),
```
Avoid calling a function in a signature definition, so move the `convert_to_tensor` inside the function body

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+"""
+
+"""
+# Custom layers
```
This should be `##`

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+
+
+class SPADE(layers.Layer):
+    def __init__(self, filters, epsilon=1e-5):
```
In general in custom layers, accept `**kwargs` in the signature and pass it to `super().__init__(**kwargs)`, which enables layer naming, etc.

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+
+class SPADE(layers.Layer):
+    def __init__(self, filters, epsilon=1e-5):
+        super(SPADE, self).__init__()
```
Python 3 style: just `super().__init__`

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+
+
+class ResBlock(layers.Layer):
+    def __init__(self, filters):
```
Likewise for the other layers

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+        )
+    )
+    if apply_norm:
+        block.add(tfa.layers.InstanceNormalization())
```
(Note: we'll add this one to core Keras soon)

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+
+
+class FeatureMatchingLoss(keras.losses.Loss):
+    def __init__(self, *args, **kwargs):
```
`*args` should not be needed, no?

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+
+
+class DiscriminatorLoss(keras.losses.Loss):
+    def __init__(self, *args, **kwargs):
```
Likwise

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+        feature_loss_coeff=10,
+        vgg_feature_loss_coeff=0.1,
+        kl_divergence_loss_coeff=0.1,
+        *args,
```
`*args` not needed

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+    def __init__(
+        self,
+        image_size,
+        n_classes,
```
`num_`

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+        )
+        return patch_size, combined_model
+
+    def compile(self, gen_lr=1e-4, disc_lr=4e-4, *args, **kwargs):
```
`*args` not needed

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+            latent_vector, segmentation_map, labels, image, mean, variance
+        )
+        return {
+            "discriminator_loss": discriminator_loss,
```
This will only display the loss value for the latest batch rather than the average over the batches of the current epoch. To do the latter, use a `Mean()` metric tracker listed in `self.metrics` and `return metric.result() for metric in self.metrics` after updating the trackers above

fchollet commented on examples/generative/gaugan.py:
```diff
@@ -0,0 +1,792 @@
+        total_generator_loss = g_loss + kl_loss + vgg_loss + feature_loss
+
+        return {
+            "discriminator_loss": total_discriminator_loss,
```
Same

fchollet reviewed (commented):
Very cool implementation! Very clean and readable. I learned a lot about GauGAN 👍

sayakpaul commented:
@fchollet thank you for your valuable comments and glad that you liked the implementation. 

We have addressed the comments. 

Also, happy new year!

sayakpaul commented:
Additionally, we created this GIF for aiding the example:

![animation](https://user-images.githubusercontent.com/22957388/147866243-00719bd1-6229-4487-8349-b44852b13c0d.gif)

fchollet reviewed (approved):
Wonderful!

I added some copyedits. You can now generate the ipynb and md files. Thank you! 👍

sayakpaul commented:
@fchollet thank you. We had to increase the LOC to make the generation process work. Over to you now.

fchollet commented:
> I was testing the example locally, the title seems to be appearing two times:

Typically that's because you didn't use `autogen.py` to generate the files and you left a blank line at the top of your `.md` file. The keras.io page generation framework will add the title line automatically if the first line doesn't start with `#`.

> The nested bullet points seem to be broken:

Nested lists are not standard markdown unfortunately. To support nested lists, we need to include an extension that handles it. If you can figure out which one, it would go here: https://github.com/keras-team/keras-io/blob/master/scripts/autogen.py#L695

I see we already have `truly-sane-lists`, maybe it supports it but with a different syntax.

> There is a :::python identifier in the cells:

The shell code cells should not contain `!`. It gets added automatically.

sayakpaul commented:
@fchollet just readded the files.

fchollet reviewed (approved):
Thank you for the wonderful contribution! 👍
